plugins/pulp_docker/plugins/distributors/publish_steps.py

```diff
@@ -220,3 +222,126 @@ def process_main(self):
+        self.parent.symlink_list.append(symlink)
+        self._tag_names.add(item.name)
+
+    def finalize(self):
+        """
+        Write the Tag list file so that clients can retrieve the list of available Tags.
+        """
+        tags_path = os.path.join(self.parent.get_working_dir(), '.relative', 'tags')
+        misc.mkdir(tags_path)
+        with open(os.path.join(tags_path, 'list'), 'w') as list_file:
+            tag_data = {
+                'name': configuration.get_repo_registry_id(self.get_repo(), self.get_config()),
+                'tags': list(self._tag_names)}
+            list_file.write(json.dumps(tag_data))
+        # We don't need the tag names anymore
+        del self._tag_names
+
+
+class DockerRsyncPublisher(Publisher):
+
+    REPO_CONTENT_TYPES = [constants.IMAGE_TYPE_ID, constants.BLOB_TYPE_ID,
+                          constants.MANIFEST_TYPE_ID]
+
+    REPO_CONTENT_MODELS = [models.Blob, models.Manifest, models.Image]
+
+    def _get_predistributor(self):
+        """
+        Returns distributor which repo has to be published in before
+        publish in rsync distributor, content generator.
+
+        :return: predistributor that was configured in rsync distributor's config
+        :rtype: Distributor
+        """
+        return None
+
+    def _add_necesary_steps(self, date_filter=None, relative_symlinks=True, config=None):
+        """
+        This method adds all the steps that are needed to accomplish an RPM rsync publish. This
+        includes:
+
+        Unit Query Step - selects units associated with the repo based on the date_filter and
+                          creates relative symlinks
+        Tag Generation Step  - creates relative symlinks for for all tags and writes out list file
+        Rsync Step (content units) - rsyncs content units from /var/lib/pulp/content to remote
+                                     server
+        Rsync Step (content) - rsyncs symlinks from working directory to remote server
+
+        :param date_filter:  Q object with start and/or end dates, or None if start and end dates
+                             are not provided
+        :type date_filter:  mongoengine.Q or types.NoneType
+        :param config: distributor configuration
+        :type config: pulp.plugins.config.PluginCallConfiguration
+        :return: None
+        """
+        remote_repo_path = configuration.get_repo_relative_path(self.repo, self.config)
+
+        unit_info = {constants.IMAGE_TYPE_ID: {'extra_path': [''], 'model': models.Image},
+                     constants.MANIFEST_TYPE_ID: {'extra_path': ['manifests'],
+                                                  'model': models.Manifest},
+                     constants.BLOB_TYPE_ID: {'extra_path': ['blobs'], 'model': models.Blob},
+                     constants.TAG_TYPE_ID: {'extra_path': ['manifests'], 'model': models.Tag}}
+
+        for type in DockerRsyncPublisher.REPO_CONTENT_TYPES:
```

Comment:
So we don't shadow a builtin function, it would be better to use a different name than just ``type``. Perhaps ``unit_type`` or ``content_type``.